### ISSUE TITLE
Fix FN S3800 (`function-return-type`): on SonarLint when using array destructuring

### ISSUE THREAD

**Issue by saberduck** [SonarSource/SonarJS#2800]
We don't detect issue of S3800 in `f1` in SonarLint, but we do detect both issues in SonarQube context. 

```typescript
const f1 = () => {  // FN S3800
 
  const [counter] = return_array();

  if (counter) return counter; // number
  return "string";
};

function return_array() {
  return [0];
}

const f2 = () => { // Noncompliant S3800
  
  const counter = return_number();

  if (counter) return counter;
  return "string";
};

function return_number() {
  return 0;
}

```

It's not clear how this could happen, it seems that type inference doesn't work the same way in SonarLint, maybe because we use program created implicitly instead of explicit `tsconfig.json`. If this is the case, it should get fixed by creating and controlling `Program` instance. 

Reported on community https://community.sonarsource.com/t/some-problems-highlighted-in-sq-not-appearing-in-vscode-sonarlint/44160/5

**Comment by Wohops** [comment]
Migrated to Jira. New ticket number: JS-127
